### PR TITLE
Feature/production inference

### PR DESCRIPTION
**Pull request by chrise96**
Blur images using val.py

!time python val.py --weights weights/best.pt --data data/pano.yaml --imgsz 4000 --batch-size 2 --skip_evaluation --save_blurred_image

### CODE REVIEW

File: val.py [hunk 1]
```diff
@@ -283,16 +294,25 @@ def run(
         # Metrics
         for si, pred in enumerate(preds):
             labels = targets[targets[:, 0] == si, 1:]
-            if tagged_data:
-                tagged_labels = targets[:, -1]
-                gt_boxes = targets[:, 2:-1] / torch.tensor((width, height, width, height), device=device)
             nl, npr = labels.shape[0], pred.shape[0]  # number of labels, predictions
             path, shape = Path(paths[si]), shapes[si][0]
+            try:
+                # TODO do we want to do this everytime in the for loop??
```
Comment (chrise96) [inline val.py:300]: We can also do "path = Path(paths[0])" before this for loop. Then we create these variables once per dataloader iteration

File: val.py [hunk 2]
```diff
@@ -213,23 +224,23 @@ def run(
...
-        dataloader = create_dataloader(data[task],
-                                       imgsz,
-                                       batch_size,
-                                       stride,
-                                       single_cls,
-                                       pad=pad,
-                                       rect=rect,
-                                       workers=workers,
-                                       prefix=colorstr(f'{task}: '))[0]
+    if pt and not single_cls:  # check --weights are trained on --data
```
Comment (chrise96) [inline val.py:227]: Maybe the "training" code back, so we can see the changes for "adding blur code" better?

File: val.py [hunk 3]
```diff
@@ -431,6 +456,8 @@ def parse_opt():
     parser.add_argument('--half', action='store_true', help='use FP16 half-precision inference')
     parser.add_argument('--dnn', action='store_true', help='use OpenCV DNN for ONNX inference')
     parser.add_argument('--tagged-data', action='store_true', help='use tagged validation')
+    parser.add_argument('--production', action='store_true', help='ignore code parts for production')
```
Comment (epureanudiana) [inline val.py:459]: let's replace `production` with `skip_evaluation`

File: val.py [hunk 4]
```diff
@@ -302,108 +322,113 @@ def run(
...
                                               save_conf,
                                               shape,
-                                              file=save_dir / 'labels' / f'{path.stem}.txt',
-                                              json_file=save_dir / 'labels_tagged' / f'{path.stem}.json',
+                                              file=txt_path / f'{path.stem}.txt',
+                                              json_file=save_dir / 'labels_tagged' / f'{path.stem}.json', # TODO, does not work in Azure
                                               confusion_matrix=confusion_matrix)
                 else:
-                    save_one_txt(predn, save_conf, shape, file=save_dir / 'labels' / f'{path.stem}.txt')
+                    save_one_txt(predn, save_conf, shape, file=txt_path / f'{path.stem}.txt')
```
Comment (epureanudiana) [inline val.py:349]: this is removing the original functionality. 
I think we still want to save labels in the `labels` folder when we run simple validation

File: val.py [hunk 5]
```diff
@@ -245,10 +256,10 @@ def run(
     jdict, stats, ap, ap_class = [], [], [], []
     callbacks.run('on_val_start')
     pbar = tqdm(dataloader, desc=s, bar_format=TQDM_BAR_FORMAT)  # progress bar
-    for batch_i, (im, targets, paths, shapes) in enumerate(pbar):
+    for batch_i, (im0, im, targets, paths, shapes) in enumerate(pbar):
         callbacks.run('on_val_batch_start')
         if tagged_data:
-            confusion_matrix = TaggedConfusionMatrix(nc=nc)
+            confusion_matrix = TaggedConfusionMatrix(nc=nc) # TODO why in the for loop? Why not at line 247?
```
Comment (epureanudiana) [inline val.py:262]: because in the case of Tagged Validation, I want to recompute the matrix every iteration for every image.

File: val.py (hunk 4 above)
Comment (epureanudiana) [reply val.py:349]: ah, ignore this comment, I saw that txt_path contains labels

File: utils/dataloaders.py [hunk 6]
```diff
@@ -672,7 +672,7 @@ def __getitem__(self, index):
         else:
             # Load image
             img, (h0, w0), (h, w) = self.load_image(index)
-
+            im0 = img.copy()
```
Comment (epureanudiana) [inline utils/dataloaders.py:675]: if `im0` is just a copy of `img` which is a numpy array, why can't we use directly `img`?
Then we do not need to modify the dataloader

File: val.py [hunk 7]
```diff
@@ -302,108 +322,113 @@ def run(
             if single_cls:
                 pred[:, 5] = 0
             predn = pred.clone()
+            predntwee = pred.clone()  # TODO can we use predn, or are these values changed?
```
Comment (epureanudiana) [inline val.py:325]: I checked and `scale_boxes` modifies `predn`, so do not use it. 

But can you rename it to something more meaningful?

File: val.py [hunk 8]
```diff
@@ -283,16 +294,25 @@ def run(
         # Metrics
         for si, pred in enumerate(preds):
             labels = targets[targets[:, 0] == si, 1:]
-            if tagged_data:
```
Comment (epureanudiana) [inline val.py:286]: why did you remove this?

File: val.py [hunk 9]
```diff
@@ -302,108 +322,113 @@ def run(
...
             scale_boxes(im[si].shape[1:], predn[:, :4], shape, shapes[si][1])  # native-space pred
 
             # Evaluate
-            if nl:
-                tbox = xywh2xyxy(labels[:, 1:5])  # target boxes
-                scale_boxes(im[si].shape[1:], tbox, shape, shapes[si][1])  # native-space labels
-                labelsn = torch.cat((labels[:, 0:1], tbox), 1)  # native-space labels
-                correct = process_batch(predn, labelsn, iouv)
-                if plots:
-                    if tagged_data:
```
Comment (epureanudiana) [inline val.py:314]: why is this removed?

File: val.py [hunk 10]
```diff
@@ -415,7 +440,7 @@ def parse_opt():
     parser.add_argument('--conf-thres', type=float, default=0.001, help='confidence threshold')
     parser.add_argument('--iou-thres', type=float, default=0.6, help='NMS IoU threshold')
     parser.add_argument('--max-det', type=int, default=300, help='maximum detections per image')
-    parser.add_argument('--task', default='val', help='train, val, test, speed or study')
```
Comment (epureanudiana) [inline val.py:418]: why did you remove train? 
I think this is for running validation on training set

File: val.py [hunk 11]
```diff
@@ -173,36 +189,31 @@ def run(
         plots=True,
         callbacks=Callbacks(),
         compute_loss=None,
-        tagged_data=False):
+        tagged_data=False,
+        production=False,
+        save_blurred_image=False):
     # Initialize/load model and set device
-    training = model is not None
-    if training:  # called by train.py
```
Comment (sdavrieux) [inline val.py:179]: How come we don't need this anymore?
